tsuru/resource_tsuru_app.go

```diff
@@ -0,0 +1,350 @@
+	}
+
+	if !restart {
+		app.NoRestart = true
+	}
+
+	resp, err := provider.TsuruClient.AppApi.AppUpdate(ctx, name, app)
+	if err != nil {
+		return diag.Errorf("unable to update app %s: %v", name, err)
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		return diag.Errorf("unable to update app %s: error code %d", name, resp.StatusCode)
+	}
+
+	return nil
+}
+
+func resourceTsuruApplicationRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	provider := meta.(*tsuruProvider)
+	name := d.Get("name").(string)
+
+	app, _, err := provider.TsuruClient.AppApi.AppGet(ctx, name)
+	if err != nil {
+		return diag.Errorf("unable to create app %s: %v", app.Name, err)
+	}
+
+	d.Set("name", name)
+	d.Set("platform", app.Platform)
+	d.Set("pool", app.Pool)
+	d.Set("plan", app.Plan.Name)
+	d.Set("team_owner", app.TeamOwner)
+
+	if app.Description != "" {
+		d.Set("description", app.Description)
+	}
+
+	if len(app.Tags) > 0 {
+		tags := []interface{}{}
+		for _, item := range app.Tags {
+			tags = append(tags, item)
+		}
+		d.Set("tags", tags)
+	}
+
+	return nil
+}
+
+func resourceTsuruApplicationDelete(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	provider := meta.(*tsuruProvider)
+	name := d.Get("name").(string)
+
+	resp, err := provider.TsuruClient.AppApi.AppDelete(ctx, name)
+	if err != nil {
+		return diag.Errorf("unable to delete app %s: %v", name, err)
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		return diag.Errorf("unable to delete app %s: error code %d", name, resp.StatusCode)
+	}
+
+	return nil
+}
+
+func resourceTsuruApplicationImport(ctx context.Context, d *schema.ResourceData, meta interface{}) ([]*schema.ResourceData, error) {
+	provider := meta.(*tsuruProvider)
+
+	app, _, err := provider.TsuruClient.AppApi.AppGet(ctx, d.Id())
+	if err != nil {
+		return nil, err
+	}
+
+	d.Set("name", app.Name)
+	d.SetId(app.Name)
+
+	return []*schema.ResourceData{d}, nil
+}
+
+func metadataFromResourceData(meta interface{}) *tsuru_client.Metadata {
+	m := meta.([]interface{})
+	if len(m) == 0 || m[0] == nil {
+		return nil
+	}
+
+	cm := tsuru_client.Metadata{}
+
+	metadataMap := m[0].(map[string]interface{})
+	if v, ok := metadataMap["labels"]; ok && len(v.(map[string]interface{})) > 0 {
+		l := v.(map[string]interface{})
+
+		for key, value := range l {
+			cm.Labels = append(cm.Labels, tsuru_client.MetadataItem{Name: key, Value: value.(string)})
+		}
+	}
+
+	if v, ok := metadataMap["annotations"]; ok && len(v.(map[string]interface{})) > 0 {
+		l := v.(map[string]interface{})
+		for key, value := range l {
+			cm.Annotations = append(cm.Annotations, tsuru_client.MetadataItem{Name: key, Value: value.(string)})
+		}
+	}
+
+	if len(cm.Labels) == 0 && len(cm.Annotations) == 0 {
+		return nil
+	}
+
+	return &cm
+}
+
+func validPlatform(ctx context.Context, provider *tsuruProvider, platform string) error {
+	platforms, _, err := provider.TsuruClient.PlatformApi.PlatformList(ctx)
```

Comment:
this kind of validation must be in server-side

Reply:
I happens on the server side too. We just return a friendly error listing what is supported.